analyzer/repo-cloner-plugin/src/main/java/eu/fasten/analyzer/repoclonerplugin/RepoClonerPlugin.java

```diff
@@ -116,37 +116,94 @@ public void consume(String record) {
             String product = null;
             if (artifact != null && !artifact.isEmpty()
                     && group != null && !group.isEmpty()) {
-                product = group + ":" + artifact + ((version == null) ? "" : ":" + version);
+                product = group + ":" + artifact + (version.isEmpty() ? "" : ":" + version);
             }
             outputPath = File.separator
                     + (artifact == null ? "" : artifact.charAt(0) + File.separator)
                     + artifact + File.separator
                     + (product == null ? "unknown" : product.replace(":", "_")) + ".json";
-            var repoUrl = json.optString("repoUrl");
-            if (repoUrl != null && !repoUrl.isEmpty()) {
-                try {
-                    var gitCloner = new GitCloner(baseDir);
-                    cloneRepo(repoUrl, gitCloner);
-                } catch (GitAPIException | IOException e) {
-                    logger.error("Error cloning repository"
-                            + (product == null ? "" : "for '" + product + "'")
-                            + " from " + repoUrl, e);
-                    this.pluginError = e;
-                    return;
+            var repoUrl = json.optString("repoUrl").replaceAll("[\\n\\t ]", "");;
+            if (!repoUrl.isEmpty()) {
+                var gitCloner = new GitCloner(baseDir);
+                var hgCloner = new HgCloner(baseDir);
+                var svnCloner = new SvnCloner(baseDir);
+                var cloned = cloneRepo(repoUrl, artifact, group, gitCloner, hgCloner, svnCloner);
+                if (!cloned) {
+                    logger.info("Could not clone repository of " + product + " from " + repoUrl);
+                    pluginError = new RuntimeException("Failed to clone repository of "
+                            + product + " from " + repoUrl);
                 }
                 if (getPluginError() == null) {
                     logger.info("Cloned repository"
-                            + (product == null ? "" : "of '" + product + "'")
+                            + (product == null ? "" : " of '" + product + "'")
                             + " from " + repoUrl + " to " + repoPath);
                 }
             } else {
                 logger.info("Repository URL not found");
             }
```

Comment:
I don't think that RuntimeException is descriptive enough in this case. You catch all exceptions within `cloneRepo([...])` method and set `cloned` to either true or false. I would suggest throwing those exceptions from the method and catch Throwable in `consume` method. This way we will have different exceptions per different repo types and they will be more descriptive, compared to only having RuntimeExceptions.

Reply:
It's a bit more complicated than that. The thing is that it is sometimes hard to tell which kind of repository is it by the URL. For example, BitBucket hosts both Git and Mercurial repositories, and their URLs would be exactly the same. So if one of the cloners (GitCloner for example) gives an error, it doesn't mean that we can't clone the repository at all. It means that it might also be simple not Git repository and we need to try all other cloners.
I am planning to do some refactoring and improve error handling in the next version of RepoCloner.

Reply:
Fair enough